plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/persistence/LibraryClasspathContainerSerializer.java

```diff
@@ -120,22 +122,23 @@ public LibraryClasspathContainer loadContainer(IJavaProject javaProject, IPath c
 
   public void saveLibraryIds(IJavaProject javaProject, List<String> libraryIds)
       throws CoreException, IOException {
-    File stateFile = getContainerStateFile(javaProject, CONTAINER_LIBRARY_LIST_FILE_ID, true);
-    if (stateFile == null) {
-      logger.warning("Master libraries file cannot be created, save failed");
+    File librariesFile = getContainerStateFile(javaProject, CONTAINER_LIBRARY_LIST_FILE_ID, true);
+    if (librariesFile == null) {
+      logger.warning("Master libraries file cannot be created, save failed"); //$NON-NLS-1$
       return;
     }
-    try (Writer out = Files.newBufferedWriter(stateFile.toPath(), StandardCharsets.UTF_8)) {
+    try (Writer out = Files.newBufferedWriter(librariesFile.toPath(), StandardCharsets.UTF_8)) {
       out.write(gson.toJson(libraryIds.toArray()));
     }
+    // delete the container state cache file since the library list has changed
+    stateLocationProvider.removeContainerStateFile(javaProject, CloudLibraries.MASTER_CONTAINER_ID);
```

Comment:
This may be the only place of this MASTER constant? I wonder if we can git rid of the constant to de-couple it from this class so that this class can be general to deal with any container. Or, is the class already the master container-specific?

Reply:
I'm not happy with this access here either, but it seemed the easiest solution without some bigger refactoring.  The issue here is that the file locations are controlled by an internal class, the `LibraryContainerStateLocationProvider`, so the outsiders can't know where these files.

The roles and responsibilities of the objects involved have become blurry. `BuildPath` collects library dependencies, but we access definitions via `CloudLibraries`.  `BuildPath` configures classpath entries, but resolution is done by `LibraryClasspathContainerResolverService`.  `LibraryClasspathContainerSerializer` shouldn't need to knows where to persist its data, it should just know how to persist data to a stream or at least to a provided file path.  But it needs to know this information as it's also responsible for creating the actual JDT container instance.  I think this could benefit from a rethink.

In the meantime, the only place that persists the library list is in `BuildPath#saveLibraryList()` (and a test), so I think it would make sense to move this `removeContainerState()` to there.

Reply:
OK, I always fail to follow what's happening with library classes. Sounds fair.

Reply:
did some refactoring, but more reorganization will need to wait for another day.